doc/Dump-Utility.md

```diff
@@ -0,0 +1,823 @@
+                                 # Again, the decision of which tables to look for and which of those are definitely supposed to have the information,
+                                   is left to the discretion of the module implementer.
+   }
+}
+Note: DB_NAME could be either of ["CONFIG_DB", "APPL_DB", "ASIC_DB", "STATE_DB", ....] or "CONFIG_FILE".
+COPP is a example which will have a entry "CONFIG_FILE". This entry is required for COPP because, the default COPP entries are read from copp_cfg.json by coppmgr and are not present in the CONFIG_DB.  
+Any extra entries which are added by the user are present in the CONFIG_DB. Hence it is required to have an extra "CONFIG_FILE" option.
+```
+
+###### Example JSON Template 1 Object:
+```
+This is an example return object from the module class for the interface name "Ethernet104" for the module "port". i.e. dump state port Ethernet104
+{
+    "CONFIG_DB": {
+        "keys": ["PORT|Ethernet104"],
+        "tables_not_found" : []
+    },
+    "APPL_DB": {
+        "keys": ["PORT_TABLE:Ethernet104"],
+        "tables_not_found" : []
+    },
+    "ASIC_DB": {
+        "keys": ["ASIC_STATE:SAI_OBJECT_TYPE_PORT:oid:0x10000000003bd", "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF:oid:0xd00000000076e"],
+        "tables_not_found": []
+    }
+    "STATE_DB": {
+        "keys": ["PORT_TABLE|Ethernet104"],
+        "tables_not_found": []
+    }
+}
+```
+
+###### Requirements on the Module Class for handling 'all' keyword
+* When the "all" argument is given by the user, get_all_args() method will be invoked and it's the responsibility of the module class to implement this.
+* get_all_args() should return a tuple of list and the list contains all the arguments for which the redis-state has to be returned.
+   * Eg: For copp, get_all_args() should return a list of trap_ids i.e. (['arp_req', 'bgpv6', 'sample_packet',..........])
+   * Eg: For port, this method will return a list of Interface names i.e (['Ethernet4', 'Ethernet8', 'Ethernet16',........])
+* An example implementation is given in the section 2.3
+* Using this information, The execute method will then be invoked for every value provided in the list.
+
+Note: VidToRid Mapping, field-value data, db filtering, printing/formatting of output to stdout, inclusion in the techsupport output are all handled in the common logic and the module need not bother about these details.
+
+### 2.2 Module Internals
+
+###### Module Example 1: PORT
+```
+class Port(Executor):
+
+    ARGS = "port_name"
+
+    def get_all_args(self, namespace):
+        all_port_names = []
+        .......  # Find and fill this list of all_port_names,
+        return (all_ports) #Eg: (['Ethernet0', 'Ethernet4', 'Ethernet8', 'Ethernet12', ....])
+
+    def execute(self, param):
+        self.template = display_template(dbs=["CONFIG_DB", "APPL_DB", "ASIC_DB"])
+        port = param[ARG_NAME]
+	ns = param["namespace"]  # namespace info, if any provided by the user
```

Comment:
In multi-asic, there could be some modules/features specific to default namespace. Ex: SNMP.
If we try to get SNMP data from CONFIG_DB, and pass a namespace, the key will not available in that namespace. so should it error out saying  table/key not found in the namespace provided?

Reply:
Yeah, that will likely happen, if not handled inside the module. 

If a module was to be programmed for SNMP and if we know, the data for SNMP will not be present in namespaces other than DEFAULT, then module implementor should take that into consideration when using MatchInfra to pull the data.